velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -0,0 +1,501 @@
+  }
+
+ private:
+  // derive from Arrow
+  inline static int256_t ReduceScaleBy(int256_t in, int32_t reduceBy) {
+    if (reduceBy == 0) {
+      // nothing to do.
+      return in;
+    }
+
+    int256_t divisor = DecimalUtil::kPowersOfTen[reduceBy];
+    DCHECK_GT(divisor, 0);
+    DCHECK_EQ(divisor % 2, 0); // multiple of 10.
+    auto result = in / divisor;
+    auto remainder = in % divisor;
+    // round up (same as BasicDecimal128::ReduceScaleBy)
+    if (abs(remainder) >= (divisor >> 1)) {
+      result += (in > 0 ? 1 : -1);
+    }
+    return result;
+  }
+};
+
+class Divide {
+ public:
+  template <typename R, typename A, typename B>
+  inline static void apply(
+      R& r,
+      const A& a,
+      const B& b,
+      uint8_t aRescale,
+      uint8_t /*bRescale*/,
+      uint8_t /* aPrecision */,
+      uint8_t /* aScale */,
+      uint8_t /* bPrecision */,
+      uint8_t /* bScale */,
+      uint8_t /* rPrecision */,
+      uint8_t /* rScale */,
+      bool* overflow) {
+    velox::sparksql::DecimalUtil::divideWithRoundUp<R, A, B>(
+        r, a, b, false, aRescale, 0, overflow);
+  }
+
+  inline static uint8_t
+  computeRescaleFactor(uint8_t fromScale, uint8_t toScale, uint8_t rScale) {
+    return rScale - fromScale + toScale;
+  }
+
+  inline static std::pair<uint8_t, uint8_t> computeResultPrecisionScale(
+      const uint8_t aPrecision,
+      const uint8_t aScale,
+      const uint8_t bPrecision,
+      const uint8_t bScale) {
+    auto scale = std::max(6, aScale + bPrecision + 1);
+    auto precision = aPrecision - aScale + bScale + scale;
+    return adjustPrecisionScale(precision, scale);
+  }
+};
+
+std::vector<std::shared_ptr<exec::FunctionSignature>>
+decimalMultiplySignature() {
+  return {exec::FunctionSignatureBuilder()
+              .integerVariable("a_precision")
+              .integerVariable("a_scale")
+              .integerVariable("b_precision")
+              .integerVariable("b_scale")
+              .integerVariable(
+                  "r_precision", "min(38, a_precision + b_precision + 1)")
+              .integerVariable(
+                  "r_scale", "a_scale") // not same with the result type
+              .returnType("DECIMAL(r_precision, r_scale)")
+              .argumentType("DECIMAL(a_precision, a_scale)")
+              .argumentType("DECIMAL(b_precision, b_scale)")
+              .build()};
+}
+
+std::vector<std::shared_ptr<exec::FunctionSignature>> decimalDivideSignature() {
+  return {
+      exec::FunctionSignatureBuilder()
+          .integerVariable("a_precision")
+          .integerVariable("a_scale")
+          .integerVariable("b_precision")
+          .integerVariable("b_scale")
+          .integerVariable(
+              "r_precision",
+              "min(38, a_precision - a_scale + b_scale + max(6, a_scale + b_precision + 1))")
+          .integerVariable(
+              "r_scale",
+              "min(37, max(6, a_scale + b_precision + 1))") // if precision is
```

Comment:
Usually, we get result type by function signature constraint, but spark decimal arithmetic is very complex, I'm not sure if we can represent the computation by string. If has, can anyone guide me?
Or can we send a function to signature, then compute the result type.
Now I compute the result type second times, it's not a good solution, and will cause function which revokes this function gets a wrong input argument type such as `makeCheckOverflow`.

And in Spark, the decimal arithmetic format is `check_overflow(a * b)`, we will use the argument type in function `apply`, so it will be normal after this function.

Reply:
Do you have some suggestions for here? @mbasmanova 
You could just review the function signature, other part will be refactored after https://github.com/facebookincubator/velox/pull/4434 merge.

Reply:
@jinchengchenghh 

> spark decimal arithmetic is very complex,

Would you share the details?

Reply:
This is spark decimal result type computation https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala#L604

I don't know how to express the `if` in string constraint

Reply:
This is the adjustPrecisionScale
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala#L171

Reply:
@jinchengchenghh Is allowPrecisionLoss a session property? I imagine we would need to define 2 separate functions, one for allowPrecisionLoss=true and another for allowPrecisionLoss=false, and choose the right kind based on the value of allowPrecisionLoss.

```
  // When `spark.sql.decimalOperations.allowPrecisionLoss` is set to true, if the precision / scale
  // needed are out of the range of available values, the scale is reduced up to 6, in order to
  // prevent the truncation of the integer part of the decimals.
  protected def allowPrecisionLoss: Boolean = SQLConf.get.decimalOperationsAllowPrecisionLoss
```

Reply:
Yes, it is default true, current implementation is based on default value.
```
  val DECIMAL_OPERATIONS_ALLOW_PREC_LOSS =
    buildConf("spark.sql.decimalOperations.allowPrecisionLoss")
      .internal()
      .doc("When true (default), establishing the result type of an arithmetic operation " +
        "happens according to Hive behavior and SQL ANSI 2011 specification, i.e. rounding the " +
        "decimal part of the result if an exact representation is not possible. Otherwise, NULL " +
        "is returned in those cases, as previously.")
      .version("2.3.1")
      .booleanConf
      .createWithDefault(true)
```
allowPrecisionLoss is a rarely used config, someone else interested in it can add the new function for allowPrecisionLoss=false, maybe in another PR.

Reply:
Do you have suggestions to generate the result type by function `adjustPrecisionScale`? I don't have much context for string constrait which then is computed by `expression::calculation::evaluate`. @mbasmanova

Reply:
@jinchengchenghh I'm not sure I understand this question. It might be better to create GitHub issue and move this discussion over there. Please, CC: @majetideepak and @karteekmurthys